File: app/src/main/java/ptml/releasing/damages/view/ReleasingDamagesSelectDamageActivity.java
package ptml.releasing.damages.view;

import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.*;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import dagger.android.support.DaggerAppCompatActivity;
import ptml.releasing.R;
import ptml.releasing.app.local.ReleasingLocal;
import ptml.releasing.damages.model.AssignedDamage;
import ptml.releasing.download_damages.model.Damage;
import ptml.releasing.download_damages.model.DamageResponse;


import javax.inject.Inject;
import java.util.ArrayList;
import java.util.List;

public class ReleasingDamagesSelectDamageActivity extends DaggerAppCompatActivity {

    @Inject
    ReleasingLocal db;


    private TextView txtSearch;
    private EditText edtSearch;
    private ListView lstDamages;
    private Button btnReturn;

    class CargoDamagesAdapter extends BaseAdapter {

        Context context;
        List<Damage> damages;
        private LayoutInflater inflater = null;

        public CargoDamagesAdapter(Context context, List<Damage> damages) {

            this.context = context;
            this.damages = damages;

            inflater = (LayoutInflater) context
                    .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        }

        @Override
        public int getCount() {
            return damages.size();
        }

        @Override
        public Object getItem(int position) {
            return damages.get(position);
        }

        @Override
        public long getItemId(int position) {
            return damages.get(position).getId();
        }

        @Override
        public View getView(final int position, View convertView, ViewGroup parent) {

            if (convertView == null) {
                convertView = inflater.inflate(R.layout.cell_cargo_select_damages, null);
            }

            TextView txtDamageName = (TextView) convertView.findViewById(R.id.CellCargoDamagesTxtDamage);
            Button btnHigh = (Button) convertView.findViewById(R.id.CellCargoDamagesBtnHigh);
            Button btnLow = (Button) convertView.findViewById(R.id.CellCargoDamagesBtnLow);

            txtDamageName.setText(damages.get(position).getDescription());

            btnHigh.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {

                    Damage d = damages.get(position);
                    DamagesActivity.currentDamages.add(
                            new AssignedDamage(
                                    d.getId(),
                                    d.getDescription(),
                                    "",
                                    1,
                                    d.getTypeContainer(),
                                    d.getPosition(),
                                    DamagesActivity.currentDamageZone + DamagesActivity.currentDamagePoint,
                                    1));
                    finish();

                }
            });

            btnLow.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {

                    Damage d = damages.get(position);
                    DamagesActivity.currentDamages.add(
                            new AssignedDamage(
                                    d.getId(),
                                    d.getDescription(),
                                    "",
                                    1,
                                    d.getTypeContainer(),
                                    d.getPosition(),
                                    DamagesActivity.currentDamageZone + DamagesActivity.currentDamagePoint,
                                    0));
                    finish();
                }
            });

            txtDamageName.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    new AlertDialog.Builder(context)
                            .setMessage(damages.get(position).getDescription())
                            .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                                public void onClick(DialogInterface dialog, int which) {
                                    // continue with delete
                                }
                            })
                            .setIcon(android.R.drawable.ic_dialog_info)
                            .show();
                }
            });


            return convertView;
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_releasing_select_damages);
        setTitle(R.string.releasing_damages_select_damage_title);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        txtSearch = (TextView) findViewById(R.id.ReleasingDamagesTxtSearch);
        edtSearch = (AutoCompleteTextView) findViewById(R.id.ReleasingDamagesEdtSearch);
        lstDamages = (ListView) findViewById(R.id.ReleasingDamagesLstDamages);
        btnReturn = (Button) findViewById(R.id.ReleasingDamagesBtnReturn);

        refreshDamages();
        setupListeners();
    }


    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if(item.getItemId() == android.R.id.home){
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }


    private void setupListeners() {

        btnReturn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                setResult(-1);
                finish();
            }
        });

        edtSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                refreshDamages();
            }
        });
    }

    private void refreshDamages() {


        String position;
        if (DamagesActivity.currentDamagePoint.equals("BBO") ||
                DamagesActivity.currentDamagePoint.equals("CBO") ||
                DamagesActivity.currentDamagePoint.equals("FBO") ||
                DamagesActivity.currentDamagePoint.equals("BOD")
        ) {
            position = "C";
        } else {
            position = "S";
        }

        //TODO Filter by position

        DamageResponse damageResponse = db.getDamages();

        CargoDamagesAdapter adapter = new CargoDamagesAdapter(this, damageResponse != null ? damageResponse.getData() : new ArrayList<>());
        lstDamages.setAdapter(adapter);
    }
}

File: app/src/main/java/ptml/releasing/damages/view/ReleasingDamagesSelectDamageActivity.kt
package ptml.releasing.damages.view

import android.app.SearchManager
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.SearchView
import androidx.lifecycle.Observer
import permissions.dispatcher.*
import ptml.releasing.BR
import ptml.releasing.R
import ptml.releasing.app.base.BaseActivity
import ptml.releasing.app.dialogs.InfoDialog
import ptml.releasing.app.exception.ErrorHandler
import ptml.releasing.app.utils.NetworkState
import ptml.releasing.app.utils.Status
import ptml.releasing.cargo_info.model.LARGE
import ptml.releasing.cargo_info.model.SMALL
import ptml.releasing.damages.model.ReleasingAssignedDamage
import ptml.releasing.damages.view_model.SelectDamageViewModel
import ptml.releasing.databinding.ActivityReleasingSelectDamagesBinding
import ptml.releasing.download_damages.model.Damage
import timber.log.Timber

@RuntimePermissions
class ReleasingDamagesSelectDamageActivity :
    BaseActivity<SelectDamageViewModel, ActivityReleasingSelectDamagesBinding>() {
    internal lateinit var adapter: CargoDamagesAdapter


    internal inner class CargoDamagesAdapter(var context: Context) : BaseAdapter() {
        var damages = mutableListOf<Damage>()

        fun setData(damagesList: List<Damage>) {
            damages.clear()
            damages.addAll(damagesList)
            notifyDataSetChanged()
        }

        override fun getCount(): Int {
            return damages.size
        }

        override fun getItem(position: Int): Any {
            return damages[position]
        }

        override fun getItemId(position: Int): Long {
            return damages[position].id?.toLong() ?: 0
        }

        @Suppress("NAME_SHADOWING")
        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            var convertView = convertView

            if (convertView == null) {
                convertView =
                    LayoutInflater.from(parent.context)
                        .inflate(R.layout.cell_cargo_select_damages, parent, false)
            }

            val txtDamageName =
                convertView?.findViewById<View>(R.id.CellCargoDamagesTxtDamage) as TextView
            val btnHigh = convertView.findViewById<View>(R.id.CellCargoDamagesBtnHigh) as Button
            val btnLow = convertView.findViewById<View>(R.id.CellCargoDamagesBtnLow) as Button

            txtDamageName.text = damages[position].description

            btnHigh.setOnClickListener {
                val d = damages[position]
                Timber.d("Damage: %s", d)
                DamagesActivity.currentDamages.add(
                    ReleasingAssignedDamage(
                        d.id ?: 0,
                        d.description,
                        "",
                        1,
                        d.typeContainer,
                        d.position,
                        DamagesActivity.currentDamageZone + DamagesActivity.currentDamagePoint,
                        LARGE
                    )
                )
                finish()
            }

            btnLow.setOnClickListener {
                val d = damages[position]
                DamagesActivity.currentDamages.add(
                    ReleasingAssignedDamage(
                        d.id ?: 0,
                        d.description,
                        "",
                        1,
                        d.typeContainer,
                        d.position,
                        DamagesActivity.currentDamageZone + DamagesActivity.currentDamagePoint,
                        SMALL
                    )
                )
                finish()
            }

            txtDamageName.setOnClickListener {
                AlertDialog.Builder(context)
                    .setMessage(damages[position].description)
                    .setPositiveButton(android.R.string.ok) { _, _ ->
                        // continue with delete
                    }
                    .setIcon(android.R.drawable.ic_dialog_info)
                    .show()
            }


            return convertView
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setTitle(R.string.releasing_damages_select_damage_title)
        showUpEnabled(true)
        adapter = CargoDamagesAdapter(this)
        binding.ReleasingDamagesLstDamages.adapter = adapter
        binding.includeEmpty.tvMessage.text = getString(R.string.empty_damages_error_message)
        binding.ReleasingDamagesLstDamages.emptyView = binding.includeEmpty.root

        setupListeners()

        viewModel.damagesList.observe(this, Observer {
            adapter.setData(it)
        })


        viewModel.networkState.observe(this, Observer {event->
            event.getContentIfNotHandled()?.let {
                if (it == NetworkState.LOADING) {
                    showLoading(
                        binding.includeProgress.root,
                        binding.includeProgress.tvMessage,
                        R.string.loading
                    )
                } else {
                    hideLoading(binding.includeProgress.root)
                }

                if (it.status == Status.FAILED) {
                    val error = ErrorHandler(this).getErrorMessage(it.throwable)
                    showLoading(binding.includeError.root, binding.includeError.tvMessage, error)
                } else {
                    hideLoading(binding.includeError.root)
                }
            }

        })


        getDataWithPermissionCheck()
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_search, menu)


        val searchMenuItem = menu?.findItem(R.id.action_search)

        val searchView = searchMenuItem?.actionView as SearchView

        val searchManager = getSystemService(Context.SEARCH_SERVICE) as SearchManager

        searchView.setSearchableInfo(searchManager.getSearchableInfo(componentName))
        searchView.queryHint = getString(R.string.search)
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {

                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                viewModel.filter(newText)
                return true
            }
        })


        return true
    }

    private fun setupListeners() {

        binding.btnBack.setOnClickListener {
            setResult(-1)
            finish()
        }

        binding.includeEmpty.btnReload.setOnClickListener {
            setResult(-1)
            finish()
        }
    }


    @NeedsPermission(android.Manifest.permission.READ_PHONE_STATE)
    fun getData() {
        viewModel.getDamages(
            imei ?: "",
            DamagesActivity.typeContainer
        )
    }

    @OnShowRationale(android.Manifest.permission.READ_PHONE_STATE)
    fun showPhoneStateRationale(request: PermissionRequest) {
        val dialogFragment = InfoDialog.newInstance(
            title = getString(R.string.allow_permission),
            message = getString(R.string.allow_phone_state_permission_msg),
            buttonText = getString(android.R.string.ok),
            listener = object : InfoDialog.InfoListener {
                override fun onConfirm() {
                    request.proceed()
                }
            })
        dialogFragment.show(supportFragmentManager, dialogFragment.javaClass.name)
    }

    @OnPermissionDenied(android.Manifest.permission.READ_PHONE_STATE)
    fun showDeniedForPhoneState() {
        notifyUser(binding.root, getString(R.string.phone_state_permission_denied))
    }

    @OnNeverAskAgain(android.Manifest.permission.READ_PHONE_STATE)
    fun neverAskForPhoneState() {
        notifyUser(binding.root, getString(R.string.phone_state_permission_never_ask))
    }


    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        onRequestPermissionsResult(requestCode, grantResults)
    }


    override fun getLayoutResourceId() = R.layout.activity_releasing_select_damages

    override fun getBindingVariable() = BR.viewModel

    override fun getViewModelClass() = SelectDamageViewModel::class.java
}
